### DISCUSSION TITLE
Split Bar Layout

### DISCUSSION THREAD

**Discussion by dotzenith** [FelixKratz/SketchyBar#221] [Q&A] [answered]
Is there a way to keep a shared background for the left and right components but leave the middle transparent?

I've attached an image of my setup with simple-bar and I was wondering if something similar could be achieved with sketchybar as well. 

![simple-bar](https://user-images.githubusercontent.com/77712443/181429080-3f1bf22b-e4f6-4450-bdfc-357aa2a28344.png)

**Comment by FelixKratz** [comment, marked as answer]
Sure, it is possible to replicate this exact look with item backgrounds and brackets.

See for example this setup I have currently running:
<img width="1512" alt="Screen Shot 2022-07-28 at 15 50 12" src="https://user-images.githubusercontent.com/22680421/181521925-1ce28e61-334c-4435-9347-b4399dccba23.png">


Simply add all the items like you would regularly, set the bar background to be transparent:
```bash
sketchybar --bar color=0x00000000
```
and add a bracket for the all left and right items. If you don't want to collect all the left and right item names manually, simply add this to the end of you `sketchybarrc`:
needs jq to be installed: `brew install jq`
```bash
ITEMS=$(sketchybar --query bar | jq -r '.items | .[]')

LEFT_ITEMS=()
RIGHT_ITEMS=()
for item in $ITEMS
do
  POS=$(sketchybar --query ${item} | jq -r '.geometry.position')
  if [ "$POS" = "left" ]; then
    LEFT_ITEMS+=(${item})
  elif [ "$POS" = "right" ]; then
    RIGHT_ITEMS+=(${item})
  fi
done

sketchybar --add braket left_items ${LEFT_ITEMS[@]} \
           --set left_items background.color=0xff000000 \
                            background.height=39 \
                            background.corner_radius=11

sketchybar --add braket right_items ${RIGHT_ITEMS[@]} \
           --set right_items background.color=0xff000000 \
                             background.height=39 \
                             background.corner_radius=11
```

Caveat:
`brackets` can not contain other `brackets`.

**Reply by dotzenith** [reply]
Lovely!! Thank you for the quick answer, it's much appreciated :)

**Comment by dotzenith** [comment]
Actually, would you mind sharing the config files you're using in the screenshot, the bar is outright just refusing to render on my side for some reason and I'm using the config from your dotfiles.

**Reply by FelixKratz** [reply]
Sure, here you go:
[sketchybar_split.zip](https://github.com/FelixKratz/SketchyBar/files/9252796/sketchybar_split.zip)

My dotfiles should be working fine... I have updated them for the exact version I am using currently.

**Reply by dotzenith** [reply]
W. Thank you so much :)

**Reply by IoannisTM** [reply]
> Sure, here you go: 
[sketchybar_split.zip](https://github.com/FelixKratz/SketchyBar/files/9252799/sketchybar_split.zip)
> 
> My dotfiles should be working fine... I have updated them for the exact version I am using currently.

any ides why I  am getting the following message with your config files?
```
 [?] Add (Group) left_items: Failed to add member '', item not found
sketchybar configuation loaded..
[1]    20279 segmentation fault  sketchybar
```

**Reply by FelixKratz** [reply]
Yes, because the bracket is created without any members. I fixed the segfault in this commit: e33c4ca539b902c59e9ecc9ddfa9ef6ae27495c5. You probably miss `jq`, thats why the configuration command is not performed correctly:
```bash
brew install jq
```

**Reply by IoannisTM** [reply]
jq is installed and working as expected
I actually uninstalled sketchy bar and redownloaded your config files just to make sure I did not messed up something.
I am not sure if the below helps
also, I do not think I really understand how this part of the script works, so ...
```
➜  ~ sketchybar
[2022-07-31 22:22:58] [?] Add (Group) left_items: Failed to add member '', item not found
[1]    35839 segmentation fault  sketchybar
/.config/sketchybar/sketchybarrc: line 81: 35862 Segmentation fault: 11  sketchybar --add braket left_items ${LEFT_ITEMS[@]} --set left_items background.color=$BAR_COLOR background.height=39 background.corner_radius=11
sketchybar configuation loaded..
➜  ~ sketchybar
[2022-07-31 22:23:20] [?] Add (Group) left_items: Failed to add member '', item not found
/.config/sketchybar/sketchybarrc: line 86: 35894 Segmentation fault: 11  sketchybar --add braket right_items ${RIGHT_ITEMS[@]} --set right_items background.color=$BAR_COLOR background.height=39 background.corner_radius=11
[1]    35870 segmentation fault  sketchybar
sketchybar configuation loaded..
```

**Reply by IoannisTM** [reply]
@dotzenith  did you managed to make it work?

@FelixKratz  So , as it is the bar appears for a split second and then goes away with the errors above.
jq seems to be working since I get the below

`sketchybar --query bar | jq -r '.items | .[]'`
`apple.logo
apple.prefs
apple.activity
apple.lock
space.1
space.2
space.3
space.4
space.5
space.6
space.7
space.8
space.9
space.10
separator
system.yabai
front_app
mail
calendar
cpu.top
cpu.percent
cpu.sys
cpu.user
left_items
right_items`

Now I guess I am missing something on the Left/Right items at this point
`ITEMS=$(sketchybar --query bar | jq -r '.items | .[]')`
`LEFT_ITEMS=()`
`RIGHT_ITEMS=()`
`for item in $ITEMS`

If I try to populate the left/right it kinda works, although I can only add one item to each, because if I try to add multiple items separated by commas it errors out, and everything is transparent. and it is kinda off.
<img width="332" alt="image" src="https://user-images.githubusercontent.com/36845463/182436991-d92ca0dd-e0fc-47db-8257-34c38d8a5385.png">

Not sure what I am doing wrong to be honest.

**Reply by FelixKratz** [reply]
If you install from master at least the crash should be gone, however, it seems the items are not properly sorted into those arrays. What prints if you perform:
```bash
sketchybar --query apple.logo
```
(after deleting the lines where the two brackets are created in sketchybarrc)

Bash arrays does not take commas as separators, rather they take spaces between quoted entries to indicate separation.

**Reply by IoannisTM** [reply]
I want to believe I deleted the proper lines :)
below the result of the query
<details>

➜  ~ sketchybar --query apple.logo
{
	"name": "apple.logo",
	"type": "item",
	"geometry": {
		"drawing": "on",
		"position": "left",
		"associated_space_mask": 0,
		"associated_display_mask": 0,
		"ignore_association": "off",
		"y_offset": 0,
		"width": -1,
		"background": {
			"drawing": "off",
			"color": "0x0",
			"border_color": "0x0",
			"border_width": 0,
			"height": 0,
			"corner_radius": 0,
			"padding_left": 10,
			"padding_right": 25,
			"y_offset": 0,
			"image": {
				"value": "(null)",
				"drawing": "off",
				"scale": 1.000000
			},
			"shadow": {
				"drawing": "off",
				"color": "0xff000000",
				"angle": 30,
				"distance": 5
			}
		}
	},
	"icon": {
		"value": "􀣺",
		"drawing": "on",
		"highlight": "off",
		"color": "0xffa6da95",
		"highlight_color": "0xff000000",
		"padding_left": 3,
		"padding_right": 3,
		"y_offset": 0,
		"font": "SF Pro:Black:16.0",
		"width": 0,
		"align": "left",
		"background": {
			"drawing": "off",
			"color": "0x0",
			"border_color": "0x0",
			"border_width": 0,
			"height": 0,
			"corner_radius": 0,
			"padding_left": 0,
			"padding_right": 0,
			"y_offset": 0,
			"image": {
				"value": "(null)",
				"drawing": "off",
				"scale": 1.000000
			},
			"shadow": {
				"drawing": "off",
				"color": "0xff000000",
				"angle": 30,
				"distance": 5
			}
		},
		"shadow": {
			"drawing": "off",
			"color": "0xff000000",
			"angle": 30,
			"distance": 5
		}
	},
	"label": {
		"value": "",
		"drawing": "off",
		"highlight": "off",
		"color": "0xffcad3f5",
		"highlight_color": "0xff000000",
		"padding_left": 3,
		"padding_right": 3,
		"y_offset": 0,
		"font": "SF Pro:Semibold:13.0",
		"width": 0,
		"align": "left",
		"background": {
			"drawing": "off",
			"color": "0x0",
			"border_color": "0x0",
			"border_width": 0,
			"height": 0,
			"corner_radius": 0,
			"padding_left": 0,
			"padding_right": 0,
			"y_offset": 0,
			"image": {
				"value": "(null)",
				"drawing": "off",
				"scale": 1.000000
			},
			"shadow": {
				"drawing": "off",
				"color": "0xff000000",
				"angle": 30,
				"distance": 5
			}
		},
		"shadow": {
			"drawing": "off",
			"color": "0xff000000",
			"angle": 30,
			"distance": 5
		}
	},
	"scripting": {
		"script": "(null)",
		"click_script": "sketchybar --set $NAME popup.drawing=toggle",
		"update_freq": 0,
		"update_mask": 0,
		"updates": "when_shown"
	},
	"bounding_rects": {
		"display-1": {
			"origin": [ 15.000000, 5.000000 ],
			"size": [ 20.000000, 39.000000 ]
		},
		"display-2": {
			"origin": [ -1497.000000, 463.000000 ],
			"size": [ 20.000000, 39.000000 ]
		}
	},
	"popup": {
		"drawing": "off",
		"horizontal": "off",
		"height": -1,
		"blur_radius": 0,
		"y_offset": 0,
		"align": "left",
		"background": {
			"drawing": "on",
			"color": "0xff181926",
			"border_color": "0xffa6da95",
			"border_width": 2,
			"height": 0,
			"corner_radius": 11,
			"padding_left": 0,
			"padding_right": 0,
			"y_offset": 0,
			"image": {
				"value": "(null)",
				"drawing": "off",
				"scale": 1.000000
			},
			"shadow": {
				"drawing": "on",
				"color": "0xff000000",
				"angle": 30,
				"distance": 5
			}
		},
		"items": [
			 "apple.prefs",
			 "apple.activity",
			 "apple.lock"
		]
	}
}</details>

**Reply by FelixKratz** [reply]
Ok, now try creating a script that performs the following task while sketchybar is running (simply copy below to a file, chmod +x it and run it):
```bash
ITEMS=$(sketchybar --query bar | jq -r '.items | .[]')

LEFT_ITEMS=()
for item in $ITEMS
do
  POS=$(sketchybar --query ${item} | jq -r '.geometry.position')
  if [ "$POS" = "left" ]; then
    LEFT_ITEMS+=(${item})
  fi
done

echo "Left items: ${LEFT_ITEMS[@]}"
```
And tell me what it prints.

**Reply by IoannisTM** [reply]
Hi, 

➜  sketchybar ./testing\ queries
Left items: apple.logo space.1 space.2 space.3 space.4 space.5 space.6 space.7 space.8 space.9 space.10 separator system.yabai front_app

btw, I am not using yabai, in case it depends on this

**Reply by FelixKratz** [reply]
Ok, now put this script in the file and execute it with sketchybar running (and still having removed the below snippet from sketchybarrc):
```bash
ITEMS=$(sketchybar --query bar | jq -r '.items | .[]')

LEFT_ITEMS=()
RIGHT_ITEMS=()
for item in $ITEMS
do
  POS=$(sketchybar --query ${item} | jq -r '.geometry.position')
  if [ "$POS" = "left" ]; then
    LEFT_ITEMS+=(${item})
  elif [ "$POS" = "right" ]; then
    RIGHT_ITEMS+=(${item})
  fi
done

sketchybar --add braket left_items ${LEFT_ITEMS[@]} \
           --set left_items background.color=0xff000000 \
                            background.height=39 \
                            background.corner_radius=11

sketchybar --add braket right_items ${RIGHT_ITEMS[@]} \
           --set right_items background.color=0xff000000 \
                             background.height=39 \
                             background.corner_radius=11
```
and report what it prints.

yabai is not required at all, it is just a simple item that displays some additional window information and changes border color.

**Reply by IoannisTM** [reply]
this actually seems to make it work.
although I can hardly find any difference between this code and the original one on your dots except from the bacground color.
<img width="755" alt="image" src="https://user-images.githubusercontent.com/36845463/182578216-5aa14e1d-2194-42fc-8aa7-e0e30f012666.png">

If i run seperate the command for left and right items I do get the below
```
➜  sketchybar ./testing\ queries\ copy
Left items: apple.logo space.1 space.2 space.3 space.4 space.5 space.6 space.7 space.8 space.9 space.10 separator system.yabai front_app
```
```
➜  sketchybar ./testing\ queries\ copy
RIGHT items: mail calendar cpu.top cpu.percent cpu.sys cpu.user
```

**Reply by FelixKratz** [reply]
What does:
```bash
sh --version
```
give on your system? On my system it is symlinked to bash actually, giving:
```bash
 # sh --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin20)
Copyright (C) 2007 Free Software Foundation, Inc.
```
maybe thats the problem.

**Reply by IoannisTM** [reply]
seems to be the same, apart from the fact that I run it on M1
➜  sketchybar sh --version
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin21)
Copyright (C) 2007 Free Software Foundation, Inc.

I am not sure if it helps but:
If i add the content to the sketchbarrc , like you initially had it, it does not render the bar (or it does for a sec and then disappears)
If I add just a line "sh ./testing_queries" , testing_queries being the same code but on a different script, it runs fine.
you can notice that it a 2 step render of the bar, since at first it loads eg 10 spaces, and then when the script runs it re-renders it back to the appropriate number

**Reply by FelixKratz** [reply]
This is a bit tricky to debug further... I noticed there is a `\` too much on the final line of the large configuration command in my sketchybarrc, maybe if you remove that it works, but it shouldn't be that.

Another issue I realized from your screenshot is that brackets where detecting their length based on the `drawing` property, but this is not sufficient for items which are associated to displays or spaces, which is why your left side it too long. I have improved this in d8ff755f3d00d6f89387f9517fb457f235049716.

I have also relaxed the timeout for the mach messages from child processes to the parent process, maybe this was a problem as well. Can you try current master? d8ff755f3d00d6f89387f9517fb457f235049716
```bash
brew uninstall sketchybar
brew install sketchybar --head
brew services restart sketchybar
```

**Reply by IoannisTM** [reply]
I think I had a " \ " less.
but now it seems to be working fine.
not sure if the left and right boxes should be the same in length but it works
<img width="761" alt="image" src="https://user-images.githubusercontent.com/36845463/182593380-f9a6945e-2644-439a-84fc-a2ef0979217d.png">

I will fiddle around with it a bit more and let you know 

Thanks a lot!!!

**Reply by FelixKratz** [reply]
Left and right boxes should be as wide as the content, so everything is working as expected now.
I will create a new release tonight, once this is done you can migrate to that again via:
```bash
brew uninstall sketchybar
brew install sketchybar
brew services restart sketchybar
```
Great to see it finally working!

**Reply by IoannisTM** [reply]
I was on and off to create my setup, but this split bar config is great.
Very nice work !!!
Thank you

**Reply by FelixKratz** [reply]
The new release is out, so you can safely return to the stable branch, and I updated the linked zip file with some tweaks.

**Reply by IoannisTM** [reply]
can confirm that it works as expected ;)
